File: seeyousun-backend/resort-reservation-service/src/main/java/com/taass/seeyousun/resortreservationservice/model/Reservation.java
package com.taass.seeyousun.resortreservationservice.model;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.taass.seeyousun.resortreservationservice.enums.PersistenceTypeEnum;
import jakarta.persistence.*;
import lombok.*;


/*
* Rappresenta una prenotazione in un singolo giorno. Può essere di mezza giornata al mattino al pomeriggio oppure giornata intera.
* */
@Entity
@Data
@AllArgsConstructor
@NoArgsConstructor
@Builder
@Table(name = "reservations")
public class Reservation {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private Integer numberOfSunbeds;

    private Integer reservedUmbrellaLine;

    private Integer reservedUmbrellaColumn;

    @ManyToOne
    @JoinColumn(name = "daily_reservation_id")
    @JsonBackReference
    @ToString.Exclude
    private DailyReservation dailyReservation;

    @Enumerated(EnumType.STRING)
    private PersistenceTypeEnum persistenceTypeEnum;

    private String userUid;

    public boolean isOverlapped(Reservation newReservation) {
        return isSamePlace(newReservation) && isSameDayPeriod(newReservation);
    }

    private boolean isSameDayPeriod(Reservation newReservation) {
        return (this.persistenceTypeEnum == PersistenceTypeEnum.FULL_DAY ||
                newReservation.persistenceTypeEnum == PersistenceTypeEnum.FULL_DAY ||
                this.persistenceTypeEnum == newReservation.persistenceTypeEnum);
    }

    public boolean isSamePlace(Reservation newReservation){
        return this.reservedUmbrellaLine.equals(newReservation.reservedUmbrellaLine) &&
                this.reservedUmbrellaColumn.equals(newReservation.reservedUmbrellaColumn);
    }
}

File: frontend-app/src/app/models/reservation.ts
export class reservation {
    resortId: number;
    resortName: string;
    date: string;
    numberOfSunbeds: number;
    persistenceTypeEnum: string;
    userUid: string;
    reservationId: number;
    dailyReservationId: number;
    reservedUmbrellaLine: number;
    reservedUmbrellaColumn: number;

    constructor(resortId: number, resortName: string, date: string, numberOfSunbeds: number, persistenceTypeEnum: string, userUid: string, reservationId: number, dailyReservationId: number, reservedUmbrellaLine: number, reservedUmbrellaColumn: number) {
        this.resortId = resortId;
        this.resortName = resortName;
        this.date = date;
        this.numberOfSunbeds = numberOfSunbeds;
        this.persistenceTypeEnum = persistenceTypeEnum;
        this.userUid = userUid;
        this.reservationId = reservationId;
        this.dailyReservationId = dailyReservationId;
        this.reservedUmbrellaLine = reservedUmbrellaLine;
        this.reservedUmbrellaColumn = reservedUmbrellaColumn;
    }
}